src/node.cc

```diff
@@ -3750,16 +3755,35 @@ void LoadEnvironment(Environment* env) {
   // who do not like how bootstrap_node.js sets up the module system but do
   // like Node's I/O bindings may want to replace 'f' with their own function.
   Local<Value> arg = env->process_object();
-  auto ret = f->Call(env->context(), Null(env->isolate()), 1, &arg);
-  // If there was an error during bootstrap then it was either handled by the
-  // FatalException handler or it's unrecoverable (e.g. max call stack
-  // exceeded). Either way, clear the stack so that the AsyncCallbackScope
-  // destructor doesn't fail on the id check.
-  // There are only two ways to have a stack size > 1: 1) the user manually
-  // called MakeCallback or 2) user awaited during bootstrap, which triggered
-  // _tickCallback().
-  if (ret.IsEmpty())
-    env->async_hooks()->clear_async_id_stack();
+
+  if (f->IsAsyncFunction()) {
+    Local<Promise> bootstrap_promise =
+        f->Call(Null(env->isolate()), 1, &arg).As<Promise>();
+
+    // bootstrap event loop
+    while (bootstrap_promise->State() == Promise::kPending) {
```

Comment:
This seems risky?

What if a commonjs module with first-tick eventloop work is loaded in the tree? Might it execute async code before load is finished?

Also, are you sure this ever returns, instead of just looping here?

Reply:
Since bootstrap code is all known code this doesn't seem like a problematic constraint to me, that the bootstrap promise always resolves. If you know a better way to handle this based on the right checks to see if work is still happening through this loop I'm all ears. It's only indefinitely pending promises that might cause these issues though - error cases all work out fine as they always reject the top-level promise (I have tested things like maximum call stack on bootstrap etc).

> What if a commonjs module with first-tick eventloop work is loaded in the tree? Might it execute async code before load is finished?

Do you have any examples of where first-tick-scheduled work is specifically needed through the bootstrap? If there are cases like this, it may make sense to delegate them to an init-based tick or similar perhaps rather.